src/arch/arm/c_traps.c

```diff
@@ -62,29 +62,54 @@ void VISIBLE NORETURN c_handle_enfp(void)
 }
 #endif /* CONFIG_HAVE_FPU */
 
-static inline void NORETURN c_handle_vm_fault(vm_fault_type_t type)
+void NORETURN vm_fault_slowpath(vm_fault_type_t type)
 {
-    NODE_LOCK_SYS;
-    c_entry_hook();
-
 #ifdef TRACK_KERNEL_ENTRIES
-    ksKernelEntry.path = Entry_VMFault;
-    ksKernelEntry.word = getRegister(NODE_STATE(ksCurThread), NextIP);
+    ksKernelEntry.is_fastpath = 0;
 #endif
-
     handleVMFaultEvent(type);
+
     restore_user_context();
     UNREACHABLE();
 }
 
+
 void VISIBLE NORETURN c_handle_data_fault(void)
 {
-    c_handle_vm_fault(seL4_DataFault);
+    NODE_LOCK_SYS;
+    c_entry_hook();
+
+#ifdef TRACK_KERNEL_ENTRIES
+    ksKernelEntry.path = Entry_VMFault;
+    ksKernelEntry.word = getRegister(NODE_STATE(ksCurThread), NextIP);
+#endif
+
+#ifdef CONFIG_EXCEPTION_FASTPATH
+    fastpath_vm_fault(seL4_DataFault);
+    UNREACHABLE();
+#else
+    vm_fault_slowpath(seL4_DataFault);
+    UNREACHABLE();
+#endif
```

Comment:
Why are these changes manually in-lined from c_handle_vm_fault? If we're already expecting c_handle_vm_fault to be in-lined then resulting binary should already have the call in-lined.

Reply:
Sorry, I'm not 100% sure what you mean by this. `vm_fault_slowpath` is what I renamed `c_handle_vm_fault` to so that I could reuse it both here and in the fastpath.

Reply:
Doing it like this avoids adding an extra function that is basically identical, but it violates the convention to have the kernel be the exact same if the config option is disabled, so maybe it's better to change it

Reply:
The function bodies of `c_handle_data_fault` and `c_handle_instruction_fault` are identical apart from vm_fault_type_t. Previously, c_handle_vm_fault would be inlined into each of them.  If you make the CONFIG_EXCEPTION_FASTPATH changes in the original `c_handle_vm_fault` you get the fastpath without duplicating the C code I think

Reply:
Oh yeah, of course. That's my bad